packages/server/src/fhir/binary.ts

```diff
@@ -41,70 +44,44 @@ async function handleBinaryWriteRequest(req: Request, res: Response): Promise<vo
     return;
   }
 
-  let binary: Binary | undefined = undefined;
   let binarySource: BinarySource = stream;
-
-  // From the spec: https://hl7.org/fhir/R4/binary.html#rest
-  //
-  // """
-  //   When binary data is written to the server (create/update - POST or PUT),
-  //   the data is accepted as is and treated as the content of a Binary,
-  //   including when the content type is "application/fhir+xml" or "application/fhir+json",
-  //   except for the special case where the content is actually a Binary resource.
-  // """
-  let binaryContentSpecialCase = false;
-
   if (contentType === ContentType.FHIR_JSON) {
     const str = await readStreamToString(stream);
+    binarySource = str;
+
     try {
       // The binary handler does *not* use Express body-parser in order to support raw binary data.
       // Therefore, we need to manually parse the body stream as JSON.
       const body = JSON.parse(str);
-      if (isResource(body) && body.resourceType === 'Binary' && body.id === id) {
+      if (isResource(body) && body.resourceType === 'Binary' && (!id || body.id === id)) {
         // Special case where the content is actually a Binary resource.
-        binary = body as Binary;
-        binaryContentSpecialCase = true;
-      } else {
-        // We have already consumed the stream, so we need to create a new one.
-        // Instead, use the original string as the source.
-        binarySource = str;
+        // From the spec: https://hl7.org/fhir/R4/binary.html#rest
+        //
+        // """
+        //   When binary data is written to the server (create/update - POST or PUT),
+        //   the data is accepted as is and treated as the content of a Binary,
+        //   including when the content type is "application/fhir+xml" or "application/fhir+json",
+        //   except for the special case where the content is actually a Binary resource.
+        // """
+        const resource = body as Binary;
+        const binary = await (create ? repo.createResource<Binary>(resource) : repo.updateResource<Binary>(resource));
+        await sendResponse(req, res, create ? created : allOk, binary);
+        return;
```

Comment:
If we're going to make this refactor, then I think this whole try/catch block needs to be inside the special case handler.  For example, as written, if `createResource` fails for some reason, we would jump to the `catch` block, and then proceed to upload the body as a binary.